feat(client): redesigned navigation

naomi-lgbt opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [ ] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [ ] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [ ] My pull request targets the `master` branch of freeCodeCamp.
- [ ] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

<!-- Feel free to add any additional description of changes below this line -->

Draft for nav redesign

ToDo:
remove language selector in footer + component

Commit:
feat: Add coderadio link to nav bar

Rebasing this against `next-i18n-client` manually.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix: Add radio to t-schema

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix(tools): Update test

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
feat(client, i18n): redesign nav again

- Menu button always displays
- Profile icon / sign in CTA always displays
- All other nav links hidden behind menu button
- Search bar hides behind menu button still on smaller screens

TODO:
- Additional CSS tweaks needed
- Update tests + snapshots

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
feat(client): CSS Tweaks

- Moves the menu box up
- Hides the background when on larger screens
- Places search bar in line with menu drop down on smaller screens

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix: unused css rule

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

moT01 committed:
fix: center logo + remove excess media queries + clean up css

moT01 committed:
fix: make links bigger + media query smaller

Commit:
fix: typos

- Revert renaming of AuthOrProfile navlinks variable
- Typo in search bar place holder

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix: tweak non-link background

Changes the non-link nav item background to use a fixed variable colour
instead of changing based on theme - this was causing contrast issues
on light theme.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

moT01 committed:
fix: rework NavLinks.js in prep for ability to toggle night mode

Commit:
fix: forum link new tab

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
feat: add theme indicator

Adds an indicator that shows a check mark for night theme and a cross
mark for default theme.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
feat: night mode toggle

Implements a functional night mode toggle - successfully changes the
user's theme setting. Button becomes disabled when user is not
signed in (based on user.username property).

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix: disabled button styling

When theme button is disabled (due to not being signed in), applies
a :disabled selector to change the style to that of the other non-link
objects.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix: adjust button layout

Modify CSS to get Menu and Sign-up buttons appropriately sized and in
line with the nav bar.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix: lint issue

Fixes the broken linting issue. Tests are still dead though.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix: chinese language link

Chinese link wasn't redirecting to `path` correctly. Resolved by
changing to english parameter.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

moT01 committed:
fix: made mobile responsive

moT01 committed:
fix: nav buttons sizing and location issues

moT01 committed:
fix: search bar highlights

Commit:
tools: update existing JEST tests

Updates the existing tests to pass with our new nav changes.
TODO:
- Add additional tests to ensure 100% coverage
- Still need to update cypress tests

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
tools: update Cypress tests

Updates the existing Cypress tests to pass.
TODO:
- Assert existence of new buttons
- Assert functionality of theme toggle
- Assert functionality of language selectors

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix(tools): unchain test calls

See if un-chaining the test calls in the Cypress test addresses the
timeout issue.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix(tools): add brief pause

Adds a 150ms pause to the cypress tests that run against the menu button
to allow for the DOM content to completely load - this should resolve
the issue with the menu button `.click` not functioning.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

Commit:
fix(tools): replace wait with beforeLoad

Removes the hard-coded wait, builds a custom helper function to detect
when the `onClick` event has hydrated in the HTML.

Thanks @ojeytonwilliams for the help!

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>

raisedadead commented on client/src/components/Header/components/NavLinks.js:
```diff
@@ -1,89 +1,166 @@
+};
+
+export class NavLinks extends Component {
+  goToLanguage(lang) {
+    const path = window.location.pathname;
+    switch (lang) {
+      case 'espanol':
+        return `${homeLocation}/espanol${path}`;
+      case 'english':
+        return `${homeLocation}${path}`;
+      case 'chinese':
+        return `${chineseHome}${path}`;
+      default:
+        return `${homeLocation}`;
+    }
+  }
```
- [ ] This needs to be created as a utility component (somehere near the createRedirects, etc.)
- [ ] The logic could be made declarative than a ever increasing list of cases.

nhcarrigan replied:
If I'm understanding correctly, I think I've applied this:

```js
const createLanguageRedirect = lang => {
  const path = window.location.pathname;
  if (lang === 'chinese') return `${chineseHome}${path}`;
  if (lang === 'english') return `${homeLocation}${path}`;
  return `${homeLocation}/${lang}${path}`;
};
```

Unless we need to subdomain another language, we shouldn't have to touch this component when we add additional languages to the client.

Commit:
feat(client, i18n): separate language redirect

Separates the language redirect into its own utility component, and
tweaks the code to be more scalable.

Signed-off-by: nhcarrigan <nhcarrigan@gmail.com>
